zerver/lib/push_notifications.py

```diff
@@ -1319,17 +1319,33 @@ def handle_remove_push_notification(user_profile_id: int, message_ids: list[int]
     # below the 4KB limit (leaving plenty of space for metadata).
     MAX_APNS_MESSAGE_IDS = 200
     truncated_message_ids = sorted(message_ids)[-MAX_APNS_MESSAGE_IDS:]
-    gcm_payload, gcm_options = get_remove_payload_gcm(user_profile, truncated_message_ids)
-    apns_payload = get_remove_payload_apns(user_profile, truncated_message_ids)
-    payload_data_to_encrypt = get_remove_payload_data_to_encrypt(
-        user_profile, truncated_message_ids
-    )
 
-    # We need to call both the legacy/non-E2EE and E2EE functions
-    # for sending mobile notifications, since we don't at this time
-    # know which mobile app version the user may be using.
-    send_push_notifications_legacy(user_profile, apns_payload, gcm_payload, gcm_options)
-    send_push_notifications(user_profile, payload_data_to_encrypt)
+    def get_payload_legacy(
+        apple_device_exists: bool, android_device_exists: bool
+    ) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
+        apns_payload = (
+            get_remove_payload_apns(user_profile, truncated_message_ids)
+            if apple_device_exists
+            else {}
+        )
+        gcm_payload, gcm_options = (
+            get_remove_payload_gcm(user_profile, truncated_message_ids)
+            if android_device_exists
+            else ({}, {})
+        )
+        return apns_payload, gcm_payload, gcm_options
+
+    def get_payload_to_encrypt() -> dict[str, Any]:
```

Comment:
Adding to my todo list to replace `dict[str, Any]` with a dataclass for the E2EE case. (I'll skip the non-E2EE payloads as we plan to drop them)